engine/crates/runtime/src/fetch.rs

```diff
@@ -60,6 +72,58 @@ impl Fetcher {
             inner: Arc::new(fetcher),
         }
     }
+
+    #[allow(unused_variables)]
+    pub async fn post<SleepFut>(
+        &self,
+        request: &FetchRequest<'_>,
+        sleep: impl Fn(Duration) -> SleepFut + Send + Sync,
+    ) -> FetchResult<FetchResponse>
+    where
+        SleepFut: std::future::Future<Output = ()> + Send,
+    {
+        let rate_limit = request
+            .rate_limiter
+            .limit(&request.subgraph_name)
+            .map_err(FetchError::RateLimit);
+
+        let mut result = rate_limit.and_then(|_| self.inner.post(request)).await;
+
+        let Some(retry_budget) = request.retry_budget else {
+            return result;
+        };
+
+        let mut counter = 0;
+
+        loop {
+            match result {
+                Ok(bytes) => {
+                    retry_budget.deposit();
+                    return Ok(bytes);
+                }
+                Err(err) => {
+                    if retry_budget.withdraw().is_ok() {
+                        let jitter = rand::random::<f64>() * 2.0;
+                        let exp_backoff = (100 * 2u64.pow(counter)) as f64;
+                        let backoff_ms = (exp_backoff * jitter).round() as u64;
+
+                        sleep(Duration::from_millis(backoff_ms)).await;
+
+                        counter += 1;
+
+                        let rate_limit = request
+                            .rate_limiter
+                            .limit(&request.subgraph_name)
```

Comment:
Just provide the `RateLimitContext` to `post`, There might be other stuff in it and `impl RateLimitContext for &str` is very ugly.

Reply:
This was surprisingly hard. The graphql/federation.rs went totally crazy with lifetime errors by passing a RateLimitContext in the FetchRequest.

Reply:
What was hard? the `subgraph_name` is a `&'ctx str` which is guaranteed to exist during the future lifetime, so seems like you just need to pass a `Box<dyn RateLimiterContext>` to the future